### ISSUE TITLE
twisted.vfs.test.test_ftp is enough to drive a sane man mad

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7212]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1264|
|Type|defect|
|Created|2005-10-09 13:57:22Z|
|Branch|https://github.com/twisted/twisted/tree/vfs-ftp-1264+2735-2|

<details><summary>Searchable metadata</summary>

```
trac-id__1264 1264
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_vfs_ftp_1264_2735_2 branches/vfs-ftp-1264+2735-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1128866242000000 1128866242000000
changetime__1195152610000000 1195152610000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__cablehead cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
FTPAdapterTestCase does some things that aren't wonderful.

Interacting with protocol-level code when testing an adapter from one
well-defined interface to another well-defined interface should be avoided. 
Having to instantiating the FTP() class should raise red flags.

Setting a bunch of attributes on the FTP instance is also borderline.  While it
works now, there's no guarantee these things will remain stable.  While these
attributes don't begin with an underscore, effort should be taken to avoid
relying on just setting them to work.  The undocumented dtpInstance is
particularly eggregious, since at least the others are mentioned in the class
doc string or strongly implied (in the case of "factory").

Calling ftp_ methods is equally unfortunate.  Which this reflects the current
implementation strategy of FTP, it is by no means guaranteed to remain stable. 
The interface guaranteed by a protocol is "dataReceived".  If bytes are going to
be rammed into one, that is the API to use.  This has the advantage of being
extremely stable, both in name, argument signature, and return signature.

Other problems which radix is in the process of fixing are the use of
"Deferred.result" - this is another nominally private attribute.  The only real
way to get the result of a Deferred is with a callback.  Also, relying on
Deferreds to have a result "synchronously" (ie, making a call, getting back a
Deferred, immediately using its result) is a mistake.  Implementations may
change so that what was once synchronous becomes asynchronous.  Tests should not
need to be rewritten in the face of this.  Testing for failures should be done
with TestCase.assertFailure().

Finally, a much saner overall strategy for these tests would be to make calls
directly onto the adapter.  It implements IFTPShell.  It *must* implemented
IFTPShell.  That is its purpose for existing.  Assert things about behavior of
methods on that interface, and all the protocol-specific code falls away.  The
tests become simpler and more obvious.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @therve|
|-|-|

I think I'll steal that. It seems that this should happen before [#10822](https://github.com/twisted/twisted/issues/10822), did I understand correctly?

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|cablehead commented|
|-|-|

Hey therve,

Thanks, you're awesome.  And yep, that's right about [#10822](https://github.com/twisted/twisted/issues/10822).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [20994]) Docstrings cleanups.

Refs [#7212](https://github.com/twisted/twisted/issues/7212).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to cablehead|
|-|-|

This is ready to review in vfs-ftp-1264+2735. It handles [#2735](https://github.com/twisted/twisted/issues/2735) too. Andy, can you have a look? Thanks.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|cablehead set owner to @therve|
|-|-|

Hey therve - bloody awesome! :)  I've brained dump thoughts while looking
through this area.  Some of the thoughts are just general vfs - ftp comments,
and not directly related to the 2 tickets in question.

Just checking - makeVFSAsync-1223-2 is the HEAD for vfs.  vfs-ftp-1264+2735 is
branched off trunk.  Since its just fixing up the current tests though, it
should be fine, it'll just need to be merged against makeVFSAsync-1223-2.
makeVFSAsync-1223-2 is now pretty different to trunk, making merging difficult.

I've planned that if makeVFSAsync-1223-2 ever got into good enough shape it
could be brought back to trunk, I will look through changes made to trunk since
that branches last merge forward, and manually ensure they are covered.

The comment for FileSystemToIFTPShellAdaptor's init says it takes a
IFileSystemContainer, but its still being passed a vfs.pathutils.Filesystem

```
    @type filesystem: L{ivfs.IFileSystemContainer}.
```

vfs.pathutils is scheduled for death.  Ideally adapter's should just take an
IFileSystemNode.   Actually this should probably be a seperate ticket. ([#2780](https://github.com/twisted/twisted/issues/2780))

I don't quite get why there's support for web2.streams, and
producers/consumers.  I'll check this with spiv.

The errors thrown by the adapter are inconsistent.  It doesn't look like
IFTPShell expects errors of any certain type, so I guess that's ok.  ivfs in
makeVFSAsync-1223-2 defines errors to be thrown by backends a little better, so
will hopefully clean this up a bit.

```
    ...
    return self.assertFailure(d, ivfs.NotFoundError) 
    ...
    return self.assertFailure(d, IOError)
```


FTPAdapterTestCase could possibly be removed now that there are direct tests
for the adapter. 

A couple of integration tests with the protocol would be nice.  This'll pick up
if IFTPShell changes - or if ftp makes undocumented assumptions about
IFTPShell.  As exarkun points out though, it would be a lot more robust to
drive these through an ftp server's dataReceived.

Alternatively - and this is what we've talked about for vfs backends a bit as
well, there could be a base suite of IFTPShell tests, which all implementations
subclass and get put through.  This gets most the benefits as integration tests
of warning about interface changes, or undocumented interface assumptions
without the gore.

Couple of typos:

```
def registerConsumer(self, consumer): 
    """ 
    Keep _trakc_ of the consumer and return a Deferred fired when finished. 
    """ 
```


```
class _FileToConsumerAdapter(object): 
    """ 
    An _adaptater_ for writing to a VFS file. 
    """ 
```

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|cablehead commented|
|-|-|

Yeah, I'm pretty sure we can drop the web2.stream stuff out of the ftp adapter ([#2781](https://github.com/twisted/twisted/issues/2781)), which makes tests a bit simpler.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

Replying to [cablehead](#note_7):
> A couple of integration tests with the protocol would be nice.  This'll pick up
> if IFTPShell changes - or if ftp makes undocumented assumptions about
> IFTPShell.  As exarkun points out though, it would be a lot more robust to
> drive these through an ftp server's dataReceived.

I started to look at this. This can be good, but it would needed some work on the ftp implementation to be cleanly feasible (especially DTP management). If you're OK, I'll open a ticket for that, but that'll delay this one even more.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|cablehead commented|
|-|-|

Ok, after a quick irc chat - a common base suite of tests for IFTPShell implementations would be great.  This will highlight omissions in the interface spec, like which errors to return etc. 

I think the list of things to do to close [#7212](https://github.com/twisted/twisted/issues/7212) and [#2735](https://github.com/twisted/twisted/issues/2735) are:

 * add tests for IFTPShellTestCase in twisted.test.test_ftp
 * import and reuse this in twisted.vfs.test.test_ftp
 * this'll probably mean some cleaning up of IFTPShell's interface
 * fix up the 3 comments noted above
 * delete FTPAdapterTestCase

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [21136]) 
 * Create IFTPShellTestCase, and use for both ftp and vfs tests
 * A bunch of cleanups in both vfs and ftp code

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to cablehead|
|-|-|

Alright, something can be checked out. I wasn't very protective on current vfs interface, mainly on errors raised: the shell needed to get consistent errors.

The branch is a bit too big to my taste. If you confirm, I'll create another ticket to merge the ftp changes first.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

twisted.vfs.test.test_sftp.SFTPAdapterTest.test_openFileTrunc failing with the branch merged into trunk at the moment, not sure if that's expected or not.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [21192]) Manage O_TRUNC in inmem backend.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to cablehead|
|-|-|

It wasn't expected, it's an effect due to a correction I made in inmem backend. That should be better now.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|cablehead set owner to @therve|
|-|-|

Hey Therve,

Apologies for delay in review :(

This looks nice!

 * No worries about the trunk vfs interface.  It's overhauled in: http://twistedmatrix.com/trac/browser/branches/makeVFSAsync-1223-2/twisted/vfs/ivfs.py

 * Should we do something similar with ftp.IReadFile and ftp.IWriteFile ?
 
 * Merging forward makeVFSAsync-1223-2 against this is starting to look hairy.  It might be easier for me to reapply the changes from that branch manually to a new branch, once this branch is merged with trunk.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

(In [21217]) Add tests for the IReadFile and IWriteFile interfaces.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set owner to cablehead|
|-|-|

No worry for the delay :).

I added tests for ftp.IReadFile and ftp.IWriteFile, back to review.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|cablehead commented|
|-|-|

Amazing :)  I'm for merging and closing these 2 tickets out.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cablehead set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve commented|
|-|-|

(In [21219]) Workaround for win32.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve set status to `closed`|
|-|-|

(In [21220]) Merge vfs-ftp-1264+2735

Author: therve
Reviewers: cablehead, exarkun
Fixes [#7212](https://github.com/twisted/twisted/issues/7212)
Fixes [#2735](https://github.com/twisted/twisted/issues/2735)

Add tests for FTP interfaces IFTPShell, IReadFile and IWriteFile. This led to
some cleanups in ftp.FTPShell, and removed several bugs in vfs ftp adapter and
inmem backend.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve commented|
|-|-|

(In [21222]) Revert [21220]: wrong commit

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set status to `reopened`|
|-|-|

I messed up in my commit.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_25"></a>|@therve commented|
|-|-|

(In [21224]) Some cleanups.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

I put it back for review, because there were a few typos in docstrings remaining.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cablehead set owner to cablehead|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|cablehead commented|
|-|-|

Couple more typo's:

vfs/adapters/test/test_ftp.py:

```
+        """
+        Create a filesystem for tests and intantiate a FTP shell.
+        """
```

vfs/adapters/ftp.py:

```
+    """
+    An adaptater for writing to a VFS file.
+    """
```

twisted/test/test_ftp.py:

```
     """
     Test functionnalities for FTP shells classes.
     """
```


```
+        """
+        openForWring should fail with L{ftp.FileNotFoundError} when trying to
+        create a file in a non-existing directory.
+        """
```

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_29"></a>|cablehead set owner to @therve|
|-|-|

I think that's it, let's merge this sucker!

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_30"></a>|cablehead set owner to @radix|
|-|-|

glyph rightly requested that it would be good for this branch to get a review from a more seasoned twisted contributor.

radix volunteered for something in the vicinity of that discussion - so assigning your way radix :) - hope that's cool.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_31"></a>|cablehead commented|
|-|-|

btw, therve, was the reason for the trunk revert just svn mis-usage?

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_32"></a>|@radix commented|
|-|-|

Ok, I'll take a look at it. I've skimmed it briefly and am slightly concerned about the API changes in the ftp code. Will discuss after dinner.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_33"></a>|@therve commented|
|-|-|

The revert happened because I merged the wrong revision of the branch, before the fixes of [21219] for win32.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve commented|
|-|-|

(In [21229]) Adjust docstrings.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_35"></a>|cablehead commented|
|-|-|

there's still the typo: openForWring in test_ftp.py

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_36"></a>|cablehead commented|
|-|-|

The only changes in API for ftp are standardizing some of the the errors to be raised. 

Tests were added for ftp.FTPShell.  As this hadn't had tests before, it highlighted omissions in IFTPShell.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cablehead set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun set owner to @therve|
|-|-|

* twisted/test/test_ftp.py
  * IFTPShellTestCase should be named something like IFTPShellTestsMixin, since it is intended to be mixed in and isn't a TestCase.
  * IFTPShellTestCase.directoryExists doesn't document how the check should report its results, nor the type of its parameter (which information is also missing from the other methods on this class).
  * The first two lines of test_removeDirectory and test_removeDirectoryOnFile seem to be testing the fixture, rather than the shell implementation.  Maybe tests such as these should be separated into test methods dedicated solely to ensuring the validity of the fixture?
  * TestConsumer.write's docstring safes _Safe_ but should probably say _Save_.
  * Similar naming issue with IWriteFileTestCase.
  * Docstrings for IWriteFileTestCase.test_{read,write} could be improved
* twisted/protocols/ftp.py
  * Is there test coverage for the changes to FTPAnonymousShell.list?  In particular, the list comprehension which does ```fPath.child``` to the ```listdir``` results.
  * two of the win32 special-cases have nice comments, but one of them just says _Workaround win32 crappiness_ and the one in FTPShell.removeFile has no comment at all.

I only skimmed the twisted/vfs/ changes.  They seemed okay, but someone (maybe me, later) should take a closer to look at them.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_39"></a>|@therve removed owner|
|-|-|

Thanks for the review!

Replying to [exarkun](#note_38):
> IFTPShellTestCase should be named something like IFTPShellTestsMixin, since it is intended to be mixed in and isn't a TestCase.

Done.

> IFTPShellTestCase.directoryExists doesn't document how the check should report its results, nor the type of its parameter (which information is also missing from the other methods on this class).

Done.

> The first two lines of test_removeDirectory and test_removeDirectoryOnFile seem to be testing the fixture, rather than the shell implementation.  Maybe tests such as these should be separated into test methods dedicated solely to ensuring the validity of the fixture?

Right, I've added 2 tests for that.

> TestConsumer.write's docstring safes _Safe_ but should probably say _Save_.

Done.

> Similar naming issue with IWriteFileTestCase.

Done.

> Docstrings for IWriteFileTestCase.test_{read,write} could be improved

I've tried something :).

> Is there test coverage for the changes to FTPAnonymousShell.list?  In particular, the list comprehension which does ```fPath.child``` to the ```listdir``` results.

I think so, see the test_list test.

> two of the win32 special-cases have nice comments, but one of them just says _Workaround win32 crappiness_ and the one in FTPShell.removeFile has no comment at all.

Right, I've improved/added comments.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_40"></a>|cablehead commented|
|-|-|

[#10822](https://github.com/twisted/twisted/issues/10822) has an overview for what's left for VFS to not suck so much.

Once this ([#7212](https://github.com/twisted/twisted/issues/7212)) ticket is merged, I'll get [#2815](https://github.com/twisted/twisted/issues/2815) ready for review as quickly as I can.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_41"></a>|@jml set owner to @therve|
|-|-|

pyflakes gives the following warnings:

```
$ pyflakes twisted/protocols/ftp.py
twisted/protocols/ftp.py:27: redefinition of unused 'grp' from line 25
twisted/protocols/ftp.py:27: redefinition of unused 'pwd' from line 25
$ pyflakes twisted/vfs/adapters/ftp.py
twisted/vfs/adapters/ftp.py:22: 'stream' imported but unused
$ pyflakes twisted/test/test_ftp.py
twisted/test/test_ftp.py:418: redefinition of fuction 'checkDownload' from line 412
twisted/test/test_ftp.py:427: redefinition of fuction 'checkDownload' from line 418
twisted/test/test_ftp.py:439: redefinition of fuction 'checkDownload' from line 427
```

The diff for this branch is 1817 lines long. This is probably too big (as I think you mention in the comments on the ticket.)

`twisted/protocols/ftp.py`:
 * The exceptions all need docstrings. Given the size of this branch, I would be happy if you filed a ticket that indicated that they need docstrings.
 * I find the names `fName`, `fEntries`, and `fPath` (in `list()`) to be confusing. It's not clear what the `f` prefix implies. Please rename these variables.
 * list() also needs a docstring.
 * `FTPShell` needs a docstring. Please add one.
 * In the comment in `openForWriting`, change `EACESS` to `EACCESS`.

`twisted/vfs/adapters/ftp.py`:
 * The XXX comment in `access()` is confusing. If it is actually XXX-worthy, then it must make clear what needs to be changed.

`twisted/test/test_ftp.py`
 * I don't particularly like the style of test mixin that you use here. However, the code is both clear and clean, and my alternative would be less so at this point, so I won't recommend changes.
 * Add a blank line between `TestConsumer` docstring and `implements`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_42"></a>|@therve commented|
|-|-|

I opened [#2855](https://github.com/twisted/twisted/issues/2855) to first check the ftp changes in.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_43"></a>|@therve commented|
|-|-|

(In [21600]) Merge forward, resolve conflicts, adapt tests to changes in trunk.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_44"></a>|@therve commented|
|-|-|

(In [21601]) Cleanups.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_45"></a>|@therve removed owner|
|-|-|

Back to review **in vfs-ftp-1264+2735-2**. The diff is only 800 lines now :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_46"></a>|@exarkun set owner to @exarkun|
|-|-|

  * twisted/vfs/test/test_ftp.py
    * ```DirectFTPAdapterTestCase``` class name and class docstring could be improved: **Direct** is only informative if you knew how the code used to work (ie, by going through the protocol layer).  All tests should directly test the unit they are concerned with, so the name/docstring should reflect something more interesting, or at least leave out this redundant information.  It might be good to say that the test case is verifying that ```FileSystemToIFTPShellAdaptor``` implements ```IFTPShell``` in a particular way (maybe mentioning if it is read-only, read-write, any other interesting properties, etc).
  * twisted/vfs/test/test_inmem.py
    * There's no module docstring.
    * ```FakeDirectory.createFile``` has logic for raising an exception if an existing file is opened exclusively, I don't see any tests for this though.  Did I miss them?
    * Similarly, I don't see tests for the ```IOError``` which that method sometimes raises.
  * twisted/vfs/adapters/ftp.py
    * the ```list``` method slices the result of ```filesystem.children``` - dropping the first two elements.  I assume these are ```.``` and ```..```.  Does the interface actually guarantee this, though?
    * there should be a ticket for doing a real implementation of the ```access``` method (and maybe also or another one for extending the general ```IFTPShell``` tests to cover this, since they don't appear to now).  The **XXX** comment in that method should point to the ticket.
    * ```_FileToConsumerAdapter.registerAdapter``` shouldn't ```assert streaming```.  I guess it should ```raise NotImplementedError``` and there should be a ticket for implementing support for non-streaming producers here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_48"></a>|@therve commented|
|-|-|

(In [21607]) Process review comments

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_49"></a>|@therve removed owner|
|-|-|

Ok that should be better.

Replying to [exarkun](#note_46):
>     * the ```list``` method slices the result of ```filesystem.children``` - dropping the first two elements.  I assume these are ```.``` and ```..```.  Does the interface actually guarantee this, though?

What interface are you talking about? The ftp interface doesn't expect this (because os.listdir doesn't give that). And the vfs interface explicitly add these two elements.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_51"></a>|@therve removed owner|
|-|-|

Put it back to review now that [#2855](https://github.com/twisted/twisted/issues/2855) has been remerged.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_52"></a>|@exarkun set owner to @therve|
|-|-|

Replying to [therve](#note_49):
> Ok that should be better.
>
> Replying to [exarkun](#note_46):
> >     * the ```list``` method slices the result of ```filesystem.children``` - dropping the first two elements.  I assume these are ```.``` and ```..```.  Does the interface actually guarantee this, though?
>
> What interface are you talking about? The ftp interface doesn't expect this (because os.listdir doesn't give that). And the vfs interface explicitly add these two elements.

I was wondering if ivfs.IFileSystemContainer.children guaranteed that the first two elements of the returned sequence would be ```.``` and ```..```.  The docstring for that method doesn't say anything about those elements.

The rest of the fixes look good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_53"></a>|@therve commented|
|-|-|

(In [21727]) Add documentation for children, cleanups.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_54"></a>|@therve removed owner|
|-|-|

Is the previous modification OK?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_55"></a>|@jml set owner to @therve|
|-|-|

The diff in [21727] looks good. I've a non-thorough review of the overall diff as well.

Put a blank line (and a docstring if you are feeling generous) between the class declaration of `FileConsumer` and the following `implements()` line in `twisted/protocols/ftp.py` and then merge away.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_56"></a>|@therve commented|
|-|-|

(In [21792]) Generous is my second name.

Refs [#7212](https://github.com/twisted/twisted/issues/7212)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_57"></a>|@therve set status to `closed`|
|-|-|

(In [21793]) Merge vfs-ftp-1264+2735-2

Author: therve
Reviewers: jml, exarkun, cablehead
Fixes [#7212](https://github.com/twisted/twisted/issues/7212)
Fixes [#2735](https://github.com/twisted/twisted/issues/2735)

Refactor tests for the vfs ftp adapter, using the new IFTPShellTestsMixin,
and fixe behavior both in the ftp adapter and in the inmem backend.

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_58"></a>|cablehead commented|
|-|-|

woo!

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
